test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs

```diff
@@ -841,6 +844,129 @@ public void ShouldCacheSignatureProvider()
 
             TestUtilities.AssertFailIfErrors(context);
         }
+
+        [Fact]
+        public void ReferenceCountingTest()
+        {
+            var context = new CompareContext($"{this}.ReferenceCountingTest");
+            var cryptoProviderFactory = new CryptoProviderFactory(CryptoProviderCacheTests.CreateCacheForTesting());
+
+            var signing = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(1, signing._ReferenceCount);
+
+            var signing2 = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(2, signing._ReferenceCount);
+            Assert.Equal(2, signing2._ReferenceCount);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing2);
+
+            Assert.Equal(1, signing._ReferenceCount);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing);
+
+            Assert.Equal(0, signing._ReferenceCount);
+        }
+
+        [Fact]
+        public void ReferenceCountingTest_NoCaching()
+        {
+            var context = new CompareContext($"{this}.ReferenceCountingTest_NoCaching");
+            var cryptoProviderFactory = new CryptoProviderFactory(CryptoProviderCacheTests.CreateCacheForTesting());
+            cryptoProviderFactory.CacheSignatureProviders = false;
+
+            var signing = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(0, signing._ReferenceCount);
+
+            var signing2 = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(0, signing._ReferenceCount);
```

Comment:
Maybe use context.AddDiff() here rather than Assert?

Reply:
using context is more consistent, fair point